cpp/include/rapidsmpf/shuffler/chunk.hpp

```diff
@@ -25,145 +27,326 @@ namespace rapidsmpf::shuffler::detail {
+     * @brief Get the data of the i-th message, as a new chunk.
+     *
+     * @param new_chunk_id The ID of the new chunk.
+     * @param i The index of the message.
+     * @param stream The CUDA stream.
+     * @return A new chunk containing the data of the i-th message.
+     * @note This will create a copy of the packed data. If there is only one message and
+     * the message is a data message, the buffers will be moved to the new chunk.
+     * Otherwise a new chunk will be created by copying data.
+     *
+     * @throws std::out_of_range if the index is out of bounds.
+     */
+    Chunk get_data(ChunkID new_chunk_id, size_t i, rmm::cuda_stream_view stream);
 
-    /// GPU data buffer of the packed `cudf::table` associated with this chunk.
-    std::unique_ptr<Buffer> gpu_data;
+    /**
+     * @brief Get the size of the metadata of the i-th message.
+     *
+     * @param i The index of the message.
+     * @return The size of the metadata of the message. Zero when the message is a
+     * control message, otherwise the size of `PackedData::metadata`.
+     */
+    inline uint32_t metadata_size(size_t i) const {
+        return i == 0 ? meta_offsets_.at(0)
+                      : meta_offsets_.at(i) - meta_offsets_.at(i - 1);
+    }
 
     /**
-     * @brief Construct a new chunk of a partition.
+     * @brief Get the size of the packed data of the i-th message.
      *
-     * @param pid The ID of the partition this chunk is part of.
-     * @param cid The ID of the chunk.
-     * @param gpu_data_size If known, the size of the gpu data buffer (in bytes).
-     * @param metadata The metadata of the packed `cudf::table` that makes up this
-     * chunk.
-     *  @param gpu_data The gpu_data of the packed `cudf::table` that makes up this
-     * chunk.
+     * @param i The index of the message.
+     * @return The size of the packed data of the message. Zero when the message is a
+     * control message, otherwise the size of `PackedData::gpu_data` of the message.
      */
-    Chunk(
-        PartID pid,
-        ChunkID cid,
-        std::size_t gpu_data_size,
-        std::unique_ptr<std::vector<uint8_t>> metadata,
-        std::unique_ptr<Buffer> gpu_data
-    );
+    inline size_t data_size(size_t i) const {
+        return i == 0 ? data_offsets_.at(0)
+                      : data_offsets_.at(i) - data_offsets_.at(i - 1);
+    }
 
     /**
-     * @brief Construct a new chunk of a partition.
+     * @brief Set the data buffer.
      *
-     * @param pid The ID of the partition this chunk is part of.
-     * @param cid The ID of the chunk.
-     * @param expected_num_chunks If not zero, the number of chunks of the partition
-     * expected to get from the sending rank. Ignored when it is zero.
+     * @param data The data buffer.
      */
-    Chunk(PartID pid, ChunkID cid, std::size_t expected_num_chunks);
+    void set_data_buffer(std::unique_ptr<Buffer> data) {
+        RAPIDSMPF_EXPECTS(data_, "buffer is already set");
```

Comment:
Sorry, had a type in my last suggestion, should have been:
```suggestion
        RAPIDSMPF_EXPECTS(!data_, "buffer is already set");
```